aiida_vasp/calcs/workchains/restart.py

```diff
@@ -169,108 +143,100 @@ def run_calculation(self):
-        self.report('workchain completed after {} iterations'.format(self.ctx.iteration))
 
-        for name, port in self.spec().outputs.iteritems():
-            if port.required and name not in self.ctx.restart_calc.out:
-                self.report('the spec specifies the output {} as required but was not an output of {}<{}>'.format(
-                    name, self._calculation_class.__name__, self.ctx.restart_calc.pk))
+        if not self.exit_status:
+            self.report('{}<{}> completed after {} iterations'.format(self.__class__.__name__, self.pid, self.ctx.iteration))
 
-            if name in self.ctx.restart_calc.out:
-                node = self.ctx.restart_calc.out[name]
-                self.out(name, self.ctx.restart_calc.out[name])
-                if self._verbose:
-                    self.report("attaching the node {}<{}> as '{}'".format(node.__class__.__name__, node.pk, name))
+            for name, port in self.spec().outputs.iteritems():
+                if port.required and name not in self.ctx.restart_calc.out:
+                    self.report('the spec specifies the output {} as required '
+                                'but was not an output of {}<{}>'.format(name, self._calculation.__name__, self.ctx.restart_calc.pk))
 
+                if name in self.ctx.restart_calc.out:
+                    node = self.ctx.restart_calc.out[name]
+                    self.out(name, self.ctx.restart_calc.out[name])
+                    if self._verbose:
+                        self.report("attaching the node {}<{}> as '{}'".format(node.__class__.__name__, node.pk, name))
+        else:
+            self.report('The calculation {}<{}> returned a non-zero exit status. '
+                        'The exit status of {} is thus set to {}'.format(self._calculation, self.pid, self.__class__.__name__,
+                                                                         self.exit_status))
         return
 
     def on_destroy(self):
         """Clean remote folders of the calculations called in the workchain if the clean_workdir input is True."""
+
         if hasattr(super(BaseRestartWorkChain, self), 'on_destroy'):
             super(BaseRestartWorkChain, self).on_destroy()  # pylint: disable=no-member
-
+            # Do not clean if we do not want to or the calculation failed
+            if self.exit_status or self.inputs.restart.clean_workdir.value is False:
+                return
         else:
             super(BaseRestartWorkChain, self).on_terminated()  # pylint: disable=no-member
-
-        if not finished_compat(self) or self.inputs.clean_workdir.value is False:
+            # Do not clean if we do not want to or the calculation failed
+            if self.exit_status or self.inputs.restart.clean_workdir.value is False:
+                return
+        # Do not clean if we do not want to or the calculation failed
+        if self.exit_status or self.inputs.restart.clean_workdir.value is False:
```

Comment:
I think this line makes the two previous copies of it superfluous

Reply:
Yes, fixed.